openlr/openlr_decoder.cpp

```diff
@@ -50,6 +69,142 @@ struct alignas(kCacheLineSize) Stats
+
+    auto const it = find_if(begin(edges), end(edges), [&edgeIt](Graph::Edge const & real) {
+      if (real.GetFeatureId() == edgeIt->GetFeatureId() && real.GetSegId() == edgeIt->GetSegId())
+        return true;
+      return false;
+    });
+
+    CHECK(it != end(edges), ());
+
+    if (2 * EdgeLength(*edgeIt) >= EdgeLength(*it))
+      *edgeIt = *it;
+    else
+      path.erase(edgeIt);
+
+    LOG(LDEBUG, ("len(e):", EdgeLength(*edgeIt), "len(*it):", EdgeLength(*it)));
+  };
+
+  expandFake(begin(path));
+  expandFake(--end(path));
+}
+
+template <typename InputIterator>
+// Returns an iterator pointing to the first edge that should not be cut off.
+InputIterator CutOffset(InputIterator start, InputIterator const stop, uint32_t const offset)
+{
+  if (offset == 0)
+    return start;
+  for (uint32_t distance = 0; start != stop; ++start)
+  {
+    auto const edgeLen = EdgeLength(*start);
+    if (distance <= offset && offset < distance + edgeLen)
+    {
+      // Throw out this edge if (offst - distance) is greater than edgeLength / 2.
+      if (2 * (offset - distance) >= edgeLen)
+        ++start;
+      break;
+    }
+    distance += edgeLen;
+  }
+
+  return start;
+}
+
+template <typename InputIterator, typename OutputIterator>
+void CopyWithoutOffsets(InputIterator const start, InputIterator const stop, OutputIterator out,
+                        uint32_t const positiveOffset, uint32_t const negativeOffset)
+{
+  auto from = start;
+  auto to = stop;
+  if (distance(start, stop) > 1)
+  {
+    from = CutOffset(start, stop, positiveOffset);
+    // |to| points past the last edge we need to take.
+    to = CutOffset(reverse_iterator<InputIterator>(stop), reverse_iterator<InputIterator>(start),
+                   negativeOffset)
+             .base();
+
+    ASSERT(from != stop, ());
+
+    // If positive and negative offets are at conflict on wheter to take or drop the same edge,
+    // we need to resolve this somehow.
+    if (from == to)
+    {
+      if (positiveOffset >= negativeOffset)
+      {
+        if (from == start)
+          ++to;
+        else
+          --from;
+      }
+      else
+      {
+        if (to == stop)
+          --from;
+        else
+          ++to;
+      }
+    }
+
+    ASSERT(from < to, ());
+  }
+
+  copy(from, to, out);
+}
+
+class MwmIdByPointGetter
+{
+public:
+  MwmIdByPointGetter(Index const & index) : m_index(&index) {}
+
+  MwmSet::MwmId operator()(m2::PointD const & p)
+  {
+    auto const & countryId = GetInfoGetter().GetRegionCountryId(p);
+    ASSERT(!countryId.empty(), ("CountryId is \"\" for point", MercatorBounds::ToLatLon(p)));
+    platform::CountryFile const countryFile(countryId);
+    return m_index->GetMwmIdByCountryFile(countryFile);
+  }
+
+private:
+  storage::CountryInfoGetter & GetInfoGetter()
+  {
+    if (!m_infoGetter)
+      m_infoGetter = storage::CountryInfoReader::CreateCountryInfoReaderObsolete(GetPlatform());
+    return *m_infoGetter;
+  }
+
+  Index const * m_index;
```

Comment:
What's the reason to store index by pointer, not reference?

Reply:
Done.